resources/build/build-utils.md

```diff
@@ -0,0 +1,587 @@
+# Using the build-utils.sh builder functions
+
+The Keyman repository is standardising on bash scripts for builds. These may
+call project-specific builders, such as `tsc` for Typescript projects, `meson`
+for our cross-platform C++ projects, `xcodebuild` on macOS and iOS projects,
+`gradle` for Android, `nmake` in our Windows builds, or worse, but each project
+should also have a `build.sh` script in its root.
+
+We have standardised on parameters and structure for `build.sh` scripts. The
+objectives are:
+
+1. to be consistent in use of script parameters across all platforms and
+   projects
+2. to be self-documenting in usage (`--help` should always tell you all you need
+   to know)
+3. for the scripts to be easily readable, coherent, and straightforward for
+   anyone involved in the project to maintain.
+
+* [Jump to API definitions](#builder-api-functions-and-variables)
+
+---
+
+# Anatomy of a build script
+
+A build script is made up of three sections:
+
+* [Prologue](#build-script-prologue)
+* [Definition](#defining-build-script-parameters)
+* [Processing actions](#build-script-actions)
+
+# Build script prologue
+
+A build script should always start with the following prologue:
+
+```bash
+#!/usr/bin/env bash
+#
+# <short description of the script purpose>
+
+set -eu
+
+## START STANDARD BUILD SCRIPT INCLUDE
+# adjust relative paths as necessary
+THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BASH_SOURCE[0]}")"
+. "$(dirname "$THIS_SCRIPT")/<relative-path-to-repo-root>/resources/build/build-utils.sh"
+## END STANDARD BUILD SCRIPT INCLUDE
+
+# any other includes, such as jq.inc.sh
+
+################################ Main script ################################
+```
+
+This prologue ensures that we have a consistent environment. Explaining each section:
+
+## Shebang
+
+```bash
+#!/usr/bin/env bash
+```
+
+We use the `/usr/bin/env` prefix to ensure that we get the right version of bash
+on macOS (installed via homebrew, rather than the obsolete system-provided one).
+This also works fine on Linux, git bash on Windows, and WSL.
+
+## Bash options (`set -eu`)
+
+We use `set -eu` throughout:
+
+* `-e` to exit on any statement failure
+* `-u` to abort on unset variable use (usually coming from typos)
```

Comment:
TODO for android/build.sh

https://github.com/keymanapp/keyman/blob/5d6ffb76551961a235193a0848781410ba40a44c/android/build.sh#L9